sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuShuffledNestedLoopJoinBase.scala

```diff
@@ -0,0 +1,155 @@
+    HashClusteredDistribution(cpuLeftKeys) :: HashClusteredDistribution(cpuRightKeys) :: Nil
+
+  override protected def doExecute(): RDD[InternalRow] = {
+    throw new UnsupportedOperationException(
+      "GpuShuffledNestedLoopJoin does not support the execute() code path.")
+  }
+
+  override def childrenCoalesceGoal: Seq[CoalesceGoal] = (joinType, buildSide) match {
+    case (FullOuter, _) => Seq(RequireSingleBatch, RequireSingleBatch)
+    case (_, GpuBuildLeft) => Seq(RequireSingleBatch, null)
+    case (_, GpuBuildRight) => Seq(null, RequireSingleBatch)
+  }
+
+  override def output: Seq[Attribute] = {
+    joinType match {
+      case _: InnerLike =>
+        left.output ++ right.output
+      case LeftOuter =>
+        left.output ++ right.output.map(_.withNullability(true))
+      case RightOuter =>
+        left.output.map(_.withNullability(true)) ++ right.output
+      case FullOuter =>
+        left.output.map(_.withNullability(true)) ++ right.output.map(_.withNullability(true))
+      case j: ExistenceJoin =>
+        left.output :+ j.exists
+      case LeftExistence(_) =>
+        left.output
+      case x =>
+        throw new IllegalArgumentException(
+          s"ShuffledNestedLoopJoin should not take $x as the JoinType")
+    }
+  }
+
+  override def doExecuteColumnar() : RDD[ColumnarBatch] = {
+    val streamTime = gpuLongMetric(STREAM_TIME)
+    val targetSizeBytes = RapidsConf.GPU_BATCH_SIZE_BYTES.get(conf)
+    val localBuildOutput: Seq[Attribute] = buildPlan.output
+    val buildTime = gpuLongMetric(BUILD_TIME)
+    val buildDataSize = gpuLongMetric(BUILD_DATA_SIZE)
+    val spillCallback = GpuMetric.makeSpillCallback(allMetrics)
+    val streamAttributes = streamedPlan.output
+    val numOutputRows = gpuLongMetric(NUM_OUTPUT_ROWS)
+    val numOutputBatches = gpuLongMetric(NUM_OUTPUT_BATCHES)
+    val opTime = gpuLongMetric(OP_TIME)
+    val joinTime = gpuLongMetric(JOIN_TIME)
+    val joinOutputRows = gpuLongMetric(JOIN_OUTPUT_ROWS)
+    val nestedLoopJoinType = joinType
+
+    // Determine which table will be first in the join and bind the references accordingly
+    // so the AST column references match the appropriate table.
+    val (firstTable, secondTable) = joinType match {
+      case RightOuter => (right, left)
+      case _ => (left, right)
+    }
+    val numFirstTableColumns = firstTable.output.size
+    val boundCondition = condition.map {
+      GpuBindReferences.bindGpuReference(_, firstTable.output ++ secondTable.output)
+    }
+
+    streamedPlan.executeColumnar().zipPartitions(buildPlan.executeColumnar()) {
+      (streamIter, buildIter) => {
+        val stIt = new CollectTimeIterator("shuffled nested loop join stream",
+          streamIter, streamTime)
+        val startTime = System.nanoTime()
+
+        withResource(ConcatAndConsumeAll.getSingleBatchWithVerification(buildIter,
+          localBuildOutput)) { builtBatch =>
+          GpuColumnVector.incRefCounts(builtBatch)
```

Comment:
This should not be here, it's getting double-closed due to the two `withResource` calls on `builtBatch`.